### PR TITLE
FIPs 36-40 changes to fio.contracts

### PR DESCRIPTION
**Pull request by edrotthoff**
This branch will be used for all changes for FIPs 36 -40 please do not merge until all FIPs are completed and verified by QA.

### CODE REVIEW

File: contracts/fio.address/fio.address.cpp [hunk 1]
```diff
@@ -669,6 +727,134 @@ namespace fioio {
...
+                namesbyname.modify(fioname_iter, _self, [&](struct fioname &a) {
+                    a.bundleeligiblecountdown = (bundleeligiblecountdown - 1);
+                });
+            } else {
+                fee_amount = fee_iter->suf_amount;
+                fio_400_assert(max_fee >= (int64_t) fee_amount, "max_fee", to_string(max_fee),
+                               "Fee exceeds supplied maximum.",
+                               ErrorMaxFeeExceeded);
+
+                //NOTE -- question here, should we always record the transfer for the fees, even when its zero,
```
Comment (ericbutz) [inline contracts/fio.address/fio.address.cpp:819]: Adding comment to make sure this gets answered by @pawmmm

File: contracts/fio.address/fio.address.cpp [hunk 2]
```diff
@@ -669,6 +727,134 @@ namespace fioio {
...
+            uint64_t fee_amount = 0;
+
+            //begin fees, bundle eligible fee logic
+            const uint128_t endpoint_hash = string_to_uint128_hash(UPDATE_ENCRYPT_KEY_ENDPOINT);
+
+            auto fees_by_endpoint = fiofees.get_index<"byendpoint"_n>();
+            auto fee_iter = fees_by_endpoint.find(endpoint_hash);
+
+            //if the fee isnt found for the endpoint, then 400 error.
+            fio_400_assert(fee_iter != fees_by_endpoint.end(), "endpoint_name", REMOVE_PUB_ADDRESS_ENDPOINT,
```
Comment (ericbutz) [inline contracts/fio.address/fio.address.cpp:797]: UPDATE_ENCRYPT_KEY_ENDPOINT ?

File: contracts/fio.address/fio.address.cpp [hunk 3]
```diff
@@ -669,6 +727,134 @@ namespace fioio {
...
+                });
+            } else {
+                fee_amount = fee_iter->suf_amount;
+                fio_400_assert(max_fee >= (int64_t) fee_amount, "max_fee", to_string(max_fee),
+                               "Fee exceeds supplied maximum.",
+                               ErrorMaxFeeExceeded);
+
+                //NOTE -- question here, should we always record the transfer for the fees, even when its zero,
+                //or should we do as this code does and not do a transaction when the fees are 0.
+                fio_fees(actor, asset(reg_amount, FIOSYMBOL), REMOVE_PUB_ADDRESS_ENDPOINT);
```
Comment (ericbutz) [inline contracts/fio.address/fio.address.cpp:821]: UPDATE_ENCRYPT_KEY_ENDPOINT

File: contracts/fio.address/fio.address.cpp [hunk 4]
```diff
@@ -669,6 +727,134 @@ namespace fioio {
...
+                        permission_level{SYSTEMACCOUNT, "active"_n},
+                        "eosio"_n,
+                        "incram"_n,
+                        std::make_tuple(actor, UPDENCRYPTKEYRAM)
+                ).send();
+            }
+
+            updhandleinf(encrypt_public_key, FIO_REQUEST_CONTENT_ENCRYPTION_PUB_KEY_DATA_DESC,fioname_iter->id,actor);
+
+            fio_400_assert(fee_iter != fees_by_endpoint.end(), "endpoint_name", REMOVE_PUB_ADDRESS_ENDPOINT,
```
Comment (ericbutz) [inline contracts/fio.address/fio.address.cpp:843]: Why is this assertion duplicated here? It is also at line 797.

Should be UPDATE_ENCRYPT_KEY_ENDPOINT

File: contracts/fio.address/fio.address.cpp [hunk 5]
```diff
@@ -144,6 +150,51 @@ namespace fioio {
 
         }
 
+        // FIP-39 begin
+        inline void updhandleinf(const string &datavalue, const string &datadesc, const uint64_t &fionameid, const name &actor) {
+            auto handleinfobynameid = handleinfo.get_index<"byfionameid"_n>();
+            auto handleinfo_iter = handleinfobynameid.find(fionameid);
+            if(handleinfo_iter == handleinfobynameid.end()){
```
Comment (ericbutz) [inline contracts/fio.address/fio.address.cpp:157]: Let's discuss this code section. Adding some notes so I don't forget:

If I am reading this code correctly it looks like:
- Check to see if the fioname has an entry in the table.
  - If not, emplace the new datavalue/datadesc
  - If yes
        - use the item found as the initial value (line 166)
        - go through the table to find if there is a entry with that datadesc already
              - If yes, update that entry
              - If no, it updates the value from line 166

Won't that overwrite the initially found item, even if it is not the same datadesc?

Need to test: 
- Add entry with datavalue1/datadesc1
- Add entry with datavalue2/datadesc2
- Confirm both entries still exist

**Review by ericbutz — CHANGES_REQUESTED** [review]

File: contracts/fio.address/fio.address.cpp (hunk 5 above)
Comment (edrotthoff) [reply contracts/fio.address/fio.address.cpp:157]: Let's do a comprehensive code walk through offline at your next opportunity.

File: contracts/fio.address/fio.address.abi [hunk 6]
```diff
@@ -792,6 +845,13 @@
       ],
       "type": "fioname"
     },
+    {
+      "name": "handleinfo",
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply contracts/fio.address/fio.address.abi:849]: Will do.

File: contracts/fio.address/fio.address.cpp [hunk 7]
```diff
@@ -23,6 +23,9 @@ namespace fioio {
         domains_table domains;
         domainsales_table domainsales;
         fionames_table fionames;
+        //FIP-39 begin
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply contracts/fio.address/fio.address.cpp:26]: Comments are to orient code merges until the project delivers. We can remove them once we have a tested and fully merged release version. If u want them out now they can definitely be removed, just let me know.

File: contracts/fio.address/fio.address.cpp (hunk 2 above)
Comment (edrotthoff) [reply contracts/fio.address/fio.address.cpp:797]: Will do

File: contracts/fio.address/fio.address.cpp (hunk 3 above)
Comment (edrotthoff) [reply contracts/fio.address/fio.address.cpp:821]: Will do.

File: contracts/fio.common/fio.common.hpp [hunk 8]
```diff
@@ -501,7 +514,9 @@ namespace fioio {
     static const uint64_t LISTDOMAINRAM = 1536; // FIOESCROW - List Domain 1140 bytes round to 512 x 3
 
     static const uint64_t BASECONTENTAMOUNT = 1000; // base amount for content on newfundsreq and obt transactions
-    
+    //FIP-38 begin
+    static const uint64_t NEWFIOCHAINACCOUNTRAM = 0;
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply contracts/fio.common/fio.common.hpp:518]: we use a unique value to permit each to be set independently. if this isn't desirable we can scrap this one and use the one suggested....let me know.

File: contracts/fio.address/fio.address.cpp (hunk 1 above)
Comment (edrotthoff) [reply contracts/fio.address/fio.address.cpp:819]: is there any action for this?

File: contracts/fio.address/fio.address.cpp (hunk 4 above)
Comment (edrotthoff) [reply contracts/fio.address/fio.address.cpp:843]: yes....vestigial code, is removed

File: contracts/fio.address/fio.address.cpp (hunk 5 above)
Comment (ericbutz) [reply contracts/fio.address/fio.address.cpp]: Discussed with Ed. I was reading the string check wrong. Also decided that we are only testing with the assumption there is a single datadesc1 type. Will not test the case above.

File: contracts/fio.common/fio.common.hpp [hunk 9]
```diff
@@ -77,6 +80,13 @@
 #define ADD_NFT_ENDPOINT "add_nft"
 #define REM_NFT_ENDPOINT "remove_nft"
 #define REM_ALL_NFTS_ENDPOINT "remove_all_nfts"
+//FIP-38 begin
+#define NEW_FIO_CHAIN_ACCOUNT_ENDPOINT "new_fio_chain_account"
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply contracts/fio.common/fio.common.hpp:84]: the name in the fip is new fio chain account. we try to keep with the fip

**Review by adsorptionenthalpy — APPROVED** [review]

File: contracts/fio.common/fio.common.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ericbutz) [reply contracts/fio.common/fio.common.hpp:518]: Will confirm with tests that the new account gets the proper RAM allocation.

File: contracts/fio.staking/fio.staking.cpp [hunk 10]
```diff
@@ -497,12 +496,30 @@ class [[eosio::contract("Staking")]]  Staking: public eosio::contract {
                 }
             }
 
-            action(
-                    permission_level{get_self(), "active"_n},
-                    SYSTEMACCOUNT,
-                    "modgenlocked"_n,
-                    std::make_tuple(actor, newperiods, newlockamount, newremaininglockamount, payouts)
-            ).send();
+            //BD-3941 begin, be sure to handle edge case where we have locks and all are in the past.
```
[the remark this answers is not in the document]
Comment (ericbutz) [reply contracts/fio.staking/fio.staking.cpp:499]: Resolved in new PR: https://github.com/fioprotocol/fio.contracts/pull/288/files

**Comment by ericbutz** [comment]
Transferred to clean PR: https://github.com/fioprotocol/fio.contracts/pull/288

Closing

**Pull request closed without merging**